src/core/qgsfeaturefilterprovider.h

```diff
@@ -50,13 +50,37 @@ class CORE_EXPORT QgsFeatureFilterProvider
 
 #endif
 
+    /**
+     * Returns TRUE if the filterFeature function is thread safe, which will lead to reliance on layer ID instead
+     * of the raw layer pointer.
+     * \since QGIS 4.0
+     */
+    virtual bool isFilterThreadSafe() const { return true; }
```

Comment:
can you mark this as deprecated? it'll be removed at the same time the non-thread-safe methods are, so it's of limited lifespan/usage